File: packages/wallet/android/src/main/java/io/capawesome/capacitorjs/plugins/wallet/WalletPlugin.java
package io.capawesome.capacitorjs.plugins.wallet;

import androidx.annotation.NonNull;
import com.getcapacitor.Logger;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import io.capawesome.capacitorjs.plugins.wallet.classes.CustomException;
import io.capawesome.capacitorjs.plugins.wallet.classes.options.SaveToGoogleWalletOptions;
import io.capawesome.capacitorjs.plugins.wallet.interfaces.EmptyCallback;

@CapacitorPlugin(name = "Wallet")
public class WalletPlugin extends Plugin {

    public static final String ERROR_UNKNOWN_ERROR = "An unknown error has occurred.";
    public static final String TAG = "WalletPlugin";

    private Wallet implementation;

    @Override
    public void load() {
        super.load();
        this.implementation = new Wallet(this);
    }

    @PluginMethod
    public void addPasses(PluginCall call) {
        rejectCallAsUnimplemented(call);
    }

    @PluginMethod
    public void canAddPasses(PluginCall call) {
        rejectCallAsUnimplemented(call);
    }

    @PluginMethod
    public void saveToGoogleWallet(PluginCall call) {
        try {
            SaveToGoogleWalletOptions options = new SaveToGoogleWalletOptions(call);
            EmptyCallback callback = new EmptyCallback() {
                @Override
                public void success() {
                    resolveCall(call);
                }

                @Override
                public void error(Exception exception) {
                    rejectCall(call, exception);
                }
            };
            implementation.saveToGoogleWallet(options, callback);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    private void rejectCall(@NonNull PluginCall call, @NonNull Exception exception) {
        String message = exception.getMessage();
        if (message == null) {
            message = ERROR_UNKNOWN_ERROR;
        }
        String code = null;
        if (exception instanceof CustomException) {
            code = ((CustomException) exception).getCode();
        }
        Logger.error(TAG, message, exception);
        call.reject(message, code);
    }

    private void rejectCallAsUnimplemented(@NonNull PluginCall call) {
        call.unimplemented("This method is not available on this platform.");
    }

    private void resolveCall(@NonNull PluginCall call) {
        call.resolve();
    }
}

File: packages/wallet/ios/Plugin/WalletPlugin.swift
import Foundation
import Capacitor

@objc(WalletPlugin)
public class WalletPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "WalletPlugin"
    public let jsName = "Wallet"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "addPasses", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "canAddPasses", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "saveToGoogleWallet", returnType: CAPPluginReturnPromise)
    ]

    public static let tag = "WalletPlugin"

    private var implementation: Wallet?

    override public func load() {
        self.implementation = Wallet(plugin: self)
    }

    @objc func addPasses(_ call: CAPPluginCall) {
        do {
            let options = try AddPassesOptions(call)
            implementation?.addPasses(options, completion: { error in
                if let error = error {
                    if let customError = error as? CustomError, case .unavailable = customError {
                        self.rejectCallAsUnavailable(call)
                    } else {
                        self.rejectCall(call, error)
                    }
                    return
                }
                self.resolveCall(call)
            })
        } catch {
            self.rejectCall(call, error)
        }
    }

    @objc func canAddPasses(_ call: CAPPluginCall) {
        implementation?.canAddPasses(completion: { result, error in
            if let error = error {
                self.rejectCall(call, error)
                return
            }
            self.resolveCall(call, result)
        })
    }

    @objc func saveToGoogleWallet(_ call: CAPPluginCall) {
        rejectCallAsUnimplemented(call)
    }

    private func rejectCall(_ call: CAPPluginCall, _ error: Error) {
        CAPLog.print("[", WalletPlugin.tag, "] ", error)
        let code = (error as? CustomError)?.code
        call.reject(error.localizedDescription, code)
    }

    private func rejectCallAsUnavailable(_ call: CAPPluginCall) {
        call.unavailable("Passes cannot be added to Apple Wallet on this device.")
    }

    private func rejectCallAsUnimplemented(_ call: CAPPluginCall) {
        call.unimplemented("This method is not available on this platform.")
    }

    private func resolveCall(_ call: CAPPluginCall) {
        call.resolve()
    }

    private func resolveCall(_ call: CAPPluginCall, _ result: Result?) {
        if let result = result?.toJSObject() as? JSObject {
            call.resolve(result)
        } else {
            call.resolve()
        }
    }
}
